File: sorting/merge_sort/merge_sort.js
function merge_sort(elements) {
    merge_sort_recursive(elements, 0, elements.length)
    return elements;
}

function merge_sort_recursive(elements, start, end) {
    if (start >= end - 1) return;

    var middle = (start + (end - start) / 2) >> 0;
    merge_sort_recursive(elements, start, middle);
    merge_sort_recursive(elements, middle, end);
    merge(elements, start, middle, end);
}

function merge(elements, start, middle, end){
    leftArray = elements.slice(start, middle);
    rightArray = elements.slice(middle, end);
    
    leftSize = leftArray.length;
    rightSize = rightArray.length;
    
    left_i = 0;
    right_i = 0;
    merged_i = start;

    while (left_i < leftSize && right_i < rightSize) {
        if (leftArray[left_i] < rightArray[right_i]) {
            elements[merged_i] = leftArray[left_i];
            left_i++;
        }
        else{
            elements[merged_i] = rightArray[right_i];
            right_i++;
        }
        merged_i++;
    }

    while (left_i < leftSize) {
        elements[merged_i] = leftArray[left_i];
        left_i ++;
        merged_i ++;
    }

    while (right_i < rightSize) {
        elements[merged_i] = rightArray[right_i];
        right_i ++;
        merged_i ++;
    }
}

const input = [3, 7, 2, 9, 0, 6]

console.log(input);
console.log(`merge_sort: ${merge_sort(input.slice())}`);


// console.log(2 + (5-2) / 2);
// console.log((2 + (5-2) / 2 >>0));
// console.log(~~(2 + (5-2) / 2));
// console.log((2 + (5-2) / 2) | 0);

File: sorting/merge_sort/merge_sort.dart
List<int> mergeSort(List<int> elements) {
  mergeSortRecursive(elements, 0, elements.length);
  return elements;
}

mergeSortRecursive(List<int> elements, int start, int end) {
  if (start >= end - 1) return;
  int middle = (start + (end - start) / 2).floor();

  mergeSortRecursive(elements, start, middle);
  mergeSortRecursive(elements, middle, end);
  merge(elements, start, middle, end);
}

void merge(List<int> elements, int start, int middle, int end) {
  List<int> leftArray = elements.sublist(start, middle);
  List<int> rightArray = elements.sublist(middle, end);

  int left_size = leftArray.length;
  int right_size = rightArray.length;

  int left_i = 0;
  int right_i = 0;
  int merged_i = start;

  while (left_i < left_size && right_i < right_size) {
    if (leftArray[left_i] < rightArray[right_i]) {
      elements[merged_i] = leftArray[left_i];
      left_i++;
    } else {
      elements[merged_i] = rightArray[right_i];
      right_i++;
    }
    merged_i++;
  }

  while (left_i < left_size) {
    elements[merged_i] = leftArray[left_i];
    left_i++;
    merged_i++;
  }

  while (right_i < right_size) {
    elements[merged_i] = rightArray[right_i];
    right_i++;
    merged_i++;
  }
}

void main() {
  List<int> input = [3, 19, 8, 2, 5, 1];
  print(input);
  print("mergeSort ${mergeSort(List<int>.from(input))}");
}
